### PR TITLE
ChainSpec trait

### PR DESCRIPTION
**Pull request by arkpar**
`ChainSpec` is now stored in `Configuration` as a trait object.
This simplifies configuring `Service` at run time with chain specs based on different native `GenesisConfig`. (e.g. polkadot and kusama)

### CODE REVIEW

File: client/chain-spec/src/lib.rs [hunk 1]
```diff
@@ -104,20 +104,48 @@
...
+	/// Returns an empty JSON object if 'properties' not defined in config
+	fn properties(&self) -> Properties;
+	/// Returns a reference to defined chain spec extensions.
+	fn extensions(&self) -> &dyn GetExtension;
+	/// Add a bootnode to the list.
+	fn add_boot_node(&mut self, addr: Multiaddr);
+	/// Return spec as JSON.
+	fn as_json(&self, raw: bool) -> Result<String, String>;
+	/// Return StorageBuilder for this spec.
+	fn as_storage_builder(&self) -> &dyn BuildStorage;
```
Comment (bkchr) [inline client/chain-spec/src/lib.rs]: Why do we need this function when we add `BuildStorage` as super trait?

Comment (arkpar) [reply client/chain-spec/src/lib.rs]: rust does not support upcasting for trait objects.

File: bin/node/inspect/src/command.rs [hunk 2]
```diff
@@ -31,15 +31,12 @@ impl InspectCmd {
...
-		mut config: &mut sc_service::config::Configuration<G, E>,
-		spec_factory: impl FnOnce(&str) -> Result<Option<sc_service::ChainSpec<G, E>>, String>,
+		mut config: &mut sc_service::config::Configuration,
+		spec_factory: impl FnOnce(&str) -> Result<Box<dyn sc_service::ChainSpec>, String>,
 		version: &sc_cli::VersionInfo,
-	) -> sc_cli::Result<()> where
-		G: sc_service::RuntimeGenesis,
-		E: sc_service::ChainSpecExtension,
-	{
+	) -> sc_cli::Result<()> where {
```
Comment (bkchr) [inline bin/node/inspect/src/command.rs]:
```suggestion
	) -> sc_cli::Result<()> {
```

File: client/chain-spec/src/chain_spec.rs [hunk 3]
```diff
@@ -306,14 +307,62 @@ impl<G: RuntimeGenesis, E: serde::Serialize> ChainSpec<G, E> {
...
+
+	fn as_json(&self, raw: bool) -> Result<String, String> {
+		ChainSpec::as_json(self, raw)
+	}
+
+	fn as_storage_builder(&self) -> &dyn BuildStorage {
+		self
+	}
+}
+
```
Comment (bkchr) [inline client/chain-spec/src/chain_spec.rs]:
```suggestion
```

File: client/chain-spec/src/chain_spec.rs [hunk 4]
```diff
@@ -306,14 +307,62 @@ impl<G: RuntimeGenesis, E: serde::Serialize> ChainSpec<G, E> {
...
+
```
Comment (bkchr) [inline client/chain-spec/src/chain_spec.rs]:
```suggestion
```

File: client/chain-spec/src/extension.rs [hunk 5]
```diff
@@ -250,6 +258,23 @@ impl<B, E> Extension for Forks<B, E> where
...
+	fn get_any(&self, t: TypeId) -> &dyn Any {
+		Extension::get_any(self, t)
+	}
+}
+
+pub fn get_extension<T: 'static>(e: &dyn GetExtension) -> Option<&T> {
+	Any::downcast_ref(GetExtension::get_any(e, TypeId::of::<T>()))
+}
+
+
```
Comment (bkchr) [inline client/chain-spec/src/extension.rs]:
```suggestion
```

File: client/chain-spec/src/extension.rs [hunk 6]
```diff
@@ -250,6 +258,23 @@ impl<B, E> Extension for Forks<B, E> where
...
+	fn get_any(&self, t: TypeId) -> &dyn Any;
+}
+
+impl <E: Extension> GetExtension for E {
+	fn get_any(&self, t: TypeId) -> &dyn Any {
+		Extension::get_any(self, t)
+	}
+}
+
+pub fn get_extension<T: 'static>(e: &dyn GetExtension) -> Option<&T> {
```
Comment (bkchr) [inline client/chain-spec/src/extension.rs]: Docs missing.

File: client/service/src/builder.rs [hunk 7]
```diff
@@ -269,16 +261,13 @@ fn new_full_parts<TBl, TRtApi, TExecDisp, TGen, TCSExt>(
 	Ok((client, backend, keystore, tasks_builder))
 }
 
-impl<TGen, TCSExt> ServiceBuilder<(), (), TGen, TCSExt, (), (), (), (), (), (), (), (), ()>
-where TGen: RuntimeGenesis, TCSExt: Extension {
+impl ServiceBuilder<(), (), (), (), (), (), (), (), (), (), ()> {
 	/// Start the service builder with a configuration.
 	pub fn new_full<TBl: BlockT, TRtApi, TExecDisp: NativeExecutionDispatch + 'static>(
-		config: Configuration<TGen, TCSExt>
+		config: Configuration
```
Comment (bkchr) [inline client/service/src/builder.rs]:
```suggestion
		config: Configuration,
```

File: client/service/src/builder.rs [hunk 8]
```diff
@@ -314,12 +303,10 @@ where TGen: RuntimeGenesis, TCSExt: Extension {
 
 	/// Start the service builder with a configuration.
 	pub fn new_light<TBl: BlockT, TRtApi, TExecDisp: NativeExecutionDispatch + 'static>(
-		config: Configuration<TGen, TCSExt>
+		config: Configuration
```
Comment (bkchr) [inline client/service/src/builder.rs]:
```suggestion
		config: Configuration,
```

File: client/service/src/builder.rs [hunk 9]
```diff
@@ -476,18 +463,18 @@ impl<TBl, TRtApi, TGen, TCSExt, TCl, TFchr, TSc, TImpQu, TFprb, TFpp, TExPool, T
...
+	) -> Result<ServiceBuilder<TBl, TRtApi, TCl, TFchr, USc, TImpQu, TFprb, TFpp,
 		TExPool, TRpc, Backend>, Error> {
 		self.with_opt_select_chain(|cfg, b| builder(cfg, b).map(Option::Some))
 	}
 
 	/// Defines which import queue to use.
 	pub fn with_import_queue<UImpQu>(
 		self,
-		builder: impl FnOnce(&Configuration<TGen, TCSExt>, Arc<TCl>, Option<TSc>, Arc<TExPool>)
+		builder: impl FnOnce(&Configuration, Arc<TCl>, Option<TSc>, Arc<TExPool>)
```
Comment (bkchr) [inline client/service/src/builder.rs]:
```suggestion
		builder: impl FnOnce(&Configuration, Arc<TCl>, Option<TSc>, Arc<TExPool>),
```

File: client/service/src/builder.rs [hunk 10]
```diff
@@ -476,18 +463,18 @@ impl<TBl, TRtApi, TGen, TCSExt, TCl, TFchr, TSc, TImpQu, TFprb, TFpp, TExPool, T
 	/// Defines which head-of-chain strategy to use.
 	pub fn with_select_chain<USc>(
 		self,
-		builder: impl FnOnce(&Configuration<TGen, TCSExt>, &Arc<Backend>) -> Result<USc, Error>
-	) -> Result<ServiceBuilder<TBl, TRtApi, TGen, TCSExt, TCl, TFchr, USc, TImpQu, TFprb, TFpp,
+		builder: impl FnOnce(&Configuration, &Arc<Backend>) -> Result<USc, Error>
```
Comment (bkchr) [inline client/service/src/builder.rs]:
```suggestion
		builder: impl FnOnce(&Configuration, &Arc<Backend>) -> Result<USc, Error>,
```

File: client/service/src/builder.rs [hunk 11]
```diff
@@ -448,9 +435,9 @@ impl<TBl, TRtApi, TGen, TCSExt, TCl, TFchr, TSc, TImpQu, TFprb, TFpp, TExPool, T
 	pub fn with_opt_select_chain<USc>(
 		self,
 		select_chain_builder: impl FnOnce(
-			&Configuration<TGen, TCSExt>, &Arc<Backend>
+			&Configuration, &Arc<Backend>
```
Comment (bkchr) [inline client/service/src/builder.rs]:
```suggestion
			&Configuration, &Arc<Backend>,
```

File: utils/frame/benchmarking-cli/src/lib.rs [hunk 12]
```diff
@@ -162,15 +160,12 @@ impl BenchmarkCmd {
...
-		mut config: &mut Configuration<G, E>,
-		spec_factory: impl FnOnce(&str) -> Result<Option<ChainSpec<G, E>>, String>,
+		mut config: &mut Configuration,
+		spec_factory: impl FnOnce(&str) -> Result<Box<dyn ChainSpec>, String>,
 		version: &VersionInfo,
-	) -> sc_cli::Result<()> where
-		G: RuntimeGenesis,
-		E: ChainSpecExtension,
-	{
+	) -> sc_cli::Result<()> where {
```
Comment (bkchr) [inline utils/frame/benchmarking-cli/src/lib.rs]:
```suggestion
	) -> sc_cli::Result<()> {
```

File: client/chain-spec/src/extension.rs [hunk 13]
```diff
@@ -250,6 +258,24 @@ impl<B, E> Extension for Forks<B, E> where
...
+impl <E: Extension> GetExtension for E {
+	fn get_any(&self, t: TypeId) -> &dyn Any {
+		Extension::get_any(self, t)
+	}
+}
+
+/// Helper function that queries an extension by type from `GetExtension`
+/// trait object.
+pub fn get_extension<T: 'static>(e: &dyn GetExtension) -> Option<&T> {
+	Any::downcast_ref(GetExtension::get_any(e, TypeId::of::<T>()))
```
[the remark this answers is not in the document]
Comment (arkpar) [reply client/chain-spec/src/extension.rs]: Trait objects can't have generic methods.

**Pull request merged**